### DISCUSSION TITLE
[RFC] Notification & Message support stacking

### DISCUSSION THREAD

**Discussion by MadCcc** [ant-design/ant-design#44359] [RFCs - archive]
## Summary

Notification and Message support stacking when there's several instances displayed at the same time.

![image](https://github.com/ant-design/ant-design/assets/27722486/150c23aa-e66d-4763-9c75-9a0dda35ee12)

## Activation

Currently when there's many Notification appears, it will be look like: 

![image](https://github.com/ant-design/ant-design/assets/27722486/4d7c488a-0f8c-47a8-8c9a-49df98b047a8)

For better arrangement and Space utilization, we decided to provide a variant to place theme together, just like what IOS did:

![image](https://github.com/ant-design/ant-design/assets/27722486/9ad99b27-f586-4071-9426-23b4b33a1d90)

## API

In `notification.useNotification(config)`, `config` provide one new prop:

| Property | Type | Default |
| --- | --- | --- |
| stack | boolean \| StackConfig | true |

`StackConfig`:

| Property | Type | Default |
| --- | --- | --- |
| threshold | number | 3 |

## Basic Example

```tsx
import React from 'react';
import { notification } from 'antd';

const App: React.FC = () => {
  const [api, contextHolder] = notification.useNotification({ stack: true });

  const openNotification = (placement: NotificationPlacement) => {
    api.info({
      message: `Notification`,
      description: 'Description',
    });
  };

  return (
    <div>
      {contextHolder}
      <button onClick={openNotification}>Show Notification</button>
    </div>
  );
}
```

### Group Rules

1. Notifications created by one instance (hook) can be grouped, and there's only one group by default.
2. When numbers of Notifications is under `maxCount`, `stack` will not work.
3. (Low Priority) Could be grouped by each Notification's `group` attribute.

### Display

1. Hidden Notifications should be virtual, which means their real width or height will not work.
2. At most will only show **two** hidden Notifications behind newest one.
4. When hovering Notification group, all Notifications should be displayed. When mouse leave, Notifications should be stacked again.

**Comment by RedJue** [comment]
LGTM👍

**Comment by Yuiai01** [comment]
是默认堆叠指定层数，还是可配置的呢

**Reply by kiner-tang** [reply]
感觉这个可以做成配置的，类似 maxLength 一样的

**Reply by Yuiai01** [reply]
对，我也是想着这个。不过多层的堆叠样式该如何处理又是一个问题。

**Reply by MadCcc** [reply]
默认超过三个就堆叠，展示上也只展示三层

**Comment by afc163** [comment]
多个分组如何区分呢？

**Reply by MadCcc** [reply]
设计上还不支持多个分组，这个 API 设计的话每个 hook 创建的实例就是一个分组

**Reply by afc163** [reply]
> 这个 API 设计的话每个 hook 创建的实例就是一个分组

App.useApp 怎么做。

**Reply by MadCcc** [reply]
`App.useApp` 因为是同一个实例，如果要支持分组的话大致应该是在呼出 Notification 的时候创建新分组：
```
api.info({ newStack: true })
```
 这个是建立在 Notification 不支持自定义排序的基础上的，如果以后支持任意分组这个 api 就要重新设计一下，所以感觉优先级不高。也许会是这样：
```
api.info({ stack: 'groupA' })
```

**Reply by afc163** [reply]
感觉保留一种就行了。

**Reply by MadCcc** [reply]
嗯，一期内容不包含分组，这个可以后续实现

**Comment by bombillazo** [comment]
Hey, why can we not setup the notification options for stack, maxCount and duration using the `<ConfigProvider notifications={{...}} >` , but must do this?:
```ts
 useEffect(() => {
    notification.config({
      duration: 5,
      maxCount: 3,
    });
  }, []);

```
